Drone/drone_base.py

```diff
@@ -0,0 +1,361 @@
+            else:
+                self.logger.error("Could not take off!")
+            
+            self.logger.info(threading.current_thread().name + ": Finished takeoff")
+            finished.set()
+
+        finished = threading.Event()
+        threading.Thread(target=takeoff_thread, name="TakeoffThread-" + str(DroneBase.cnt), 
+            args=(target_altitude, finished, stop_event)).start()
+        DroneBase.cnt += 1
+        return finished
+
+    def land(self):
+        """
+        Lands the drone on the ground
+
+        Parameters
+        ----------
+        None
+
+        Precondition:
+        ----------
+        The drone is connected and armed.
+
+        Postcondition:
+        ----------
+        The drone is on the groun (and ideally not smashed into 1000 pieces)
+        
+        Returns:
+        ----------
+        None
+        """
+        def land_thread(finished):
+            self.logger.info(threading.current_thread().name + ": Starting land")
+            while not self.vehicle.mode == VehicleMode(c.LAND_MODE):
+                self.vehicle.mode = VehicleMode(c.LAND_MODE)
+            while self.vehicle.armed:
+                pass
+
+            self.logger.info(threading.current_thread().name + ": Finished  land")
+            finished.set()
+
+        finished = threading.Event()
+        threading.Thread(target=land_thread, name="LandThread-" + str(DroneBase.cnt), 
+            args=(finished,)).start()
+        DroneBase.cnt += 1
+        return finished
+
+    def move(self, direction, distance, stop_event, velocity=c.DEFAULT_VELOCITY):
+        """
+        Moves the drone along a path.
+
+        Parameters
+        ----------
+        direction: UP, DOWN, LEFT, RIGHT, FORWARD, BACK (as defined in constants.py) 
+            The direction the drone should travel in
+        distance: Double
+            The distance in meters the drone should travel in the given direction
+        emergency_land: threading.Event()
+            Set when this movement needs to stop
+
+        Precondition:
+        ----------
+        The drone is already in the air and not moving anywhere else
+
+        Postcondition:
+        ----------
+        The drone has moved in the specified direction by the specified number of meters
+
+        Returns:
+        ----------
+        threading.Event
+        """
+
+        # Calculate duration to send velocity command based on distance and velocity
+        duration = int(distance / velocity)
+
+        # Multiply unit vector in direction by the velocity
+        vector = tuple(velocity * n for n in direction)
+
+        # Make the mavlink message
+        msg = dkw.get_velocity_message(self.vehicle.message_factory, vector)
+
+        def move_thread(msg, duration, finished, stop_event):
+            self.logger.info(threading.current_thread().name + ": Starting move")
+            # Send the message once every second
+            for _ in range(0, duration):
```

Comment:
This abuses range quit a bit, here's an example of a better timed event system:

```Python
class Timer():
    def __init__(self):
        self._event_threads = []
        self._stop_event_map = {}
        self._thread_map = {}
        self._lock = threading.Lock()
        self._num_threads_lock = threading.Lock()

        self.reset()

    def _create_event_thread(self, closure, name):
        stop_event = threading.Event()
        self._stop_event_map[name] = stop_event
        thread = threading.Thread(target=closure, daemon=True, args=(stop_event,))
        self._thread_map[name] = thread
        self._event_threads.append(thread)
        thread.start()

    @property
    def num_threads(self):
        with self._num_threads_lock as lock:
            return len(self._event_threads)

    def add_callback(self, name, when_to_call, callback, recurring=False):
        # when_to_call: call when x seconds are elapsed
        if name in self._stop_event_map:
            raise ValueError('The `name` parameter cannot be assigned the duplicate name: "{}"!'.format(name))

        if when_to_call <= self.elapsed:
            callback()
            return

        def handle_event(stop_event):
            while recurring and not stop_event.is_set():
                time.sleep(when_to_call-self.elapsed)
                event_handled = False
                while ((not event_handled) or recurring) and not stop_event.is_set():
                    if when_to_call <= self.elapsed:
                        callback()
                        event_handled = True
                        break

                    time.sleep(1e-3)

                if self.num_threads == 1 and recurring:
                    self.reset()

        self._create_event_thread(handle_event, name)

    def stop_callback(self, name):
        self._stop_event_map[name].set()

    def shutdown(self):
        for name, stop_event in self._stop_event_map.items():
            stop_event.set()
            self._thread_map[name].join()

    @property
    def elapsed(self):
        with self._lock as lock:
            return timer() - self._start

    def reset(self):
        with self._lock as lock:
            self._start = timer()
```

Reply:
This is taken from a project I'm doing rapid prototype in at the moment, so the code above is not perfect, but should get the idea of how useful a object-oriented approach to timers can be

Reply:
Also, be sure to use string.format in remaining code for this file

Reply:
Nice code! I have swapped out the old range based implementation with your timer class. I also changed the hover function, which was very similar to move().